### ISSUE TITLE
Mode4 Vehicle keeps going in circles at end of row turning

### ISSUE THREAD

**Issue by Petje81** [Courseplay/courseplay#4407]
As suggested by @pvaiko a resubmission of this problem after testing following issue [https://github.com/Courseplay/courseplay/issues/4372](url).

- Amazone Pantera (ingame modhub mod with only changed workspeed to 25 kph)
- Turn speed: 15kph
- Field speed: 25 kph
- Lone Oak Field 48 on a contract
- Courseplay 6 v311
- Created a new course
- 1 Headland on "smooth"
- "use AI turn maneuvers" deactivated

The headland is done like it should with the speed dropping to 15 kph just before the corner, then fluctuating around 8 kph going through the corner and going back to 15 kph powering out of the corner and back to field speed. So no problems there.

Turning at row end the speed drops to the set turn speed of 15 kph, sometimes 14 kph and keeps it steady at that speed, overshoots the 90 degree turn, overcorrects with a 180 degree turn and then goes into a wide turn (very wide turning angle) ending up near the green pointer (next row start), overshoots again, corrects itself and ends up starting that next row....

So in the end, if you look at the goal, the system worked in this instance ;)
Only the actions that are performed to get to that goal are a bit exaggerated.

If you need any more information and/or media from me just let me know.

**Comment by Tensuko** [comment]
I think a logfile with channel 12 and 14 (iirc) gives enough information from #4372 ...

**Comment by Petje81** [comment]
Same test with the following difference:

- Turn speed: 10 kph

On the headland turns there is the same response, same speed with the only difference being the 10 kph before the turn. Even the speed through the corner is the same.

At the end of row turns it is worse than the higher speed turning. The same behaviour as V310, double overcorrection and into a continuous circle at 9 kph.

@Tensuko I just saw your reply as i was testing the lower speed.
Will follow up with a logfile on 15 kph turn speed.

**Comment by Petje81** [comment]
[20191019_log.txt](https://github.com/Courseplay/courseplay/files/3746420/20191019_log.txt)

Here you go.

I noticed some stuff regarding a set turn radius path but i don't know if you'd like me to share this with you. I am not a programmer, do like to test stuff (it's sorta my job irl) and i don't want to be rude as i really appreciate the work you guys have done and still are doing.

**Comment by pvaiko** [comment]
This is most likely the same as #4409

**Comment by tuzlo** [comment]
I had the same issue, then I reduced th Turn speed back to default and all is well now.

**Comment by Voidheartd** [comment]
I was having this issue suddenly as well. It looks to me like the tractor is steering way too slow. That causes it to miss a turn it should easily have made.

I had to slow my tractor from 10 to 8 to stop the infinite spinning but it still wobbles through the turns because it keeps oversteering one way and then the other thanks to the slow steering.

**Comment by Tensuko** [comment]
Test again with 311.1 and activated the PPC.
I will make test when I am home later.

**Comment by Voidheartd** [comment]
I thought I had gotten that update but apparently not.

Looks like using PPC causes CP to ignore the speed settings. It went 7mph during the turn and when backing up. That's also the tool work speed so that may be related.

Turning with PPC off works fine though so that issue is solved.

**Comment by Tensuko** [comment]
How can the Issue be solved, when turning off the fix solve it ?
Turn tests later..... yay.....

**Comment by pvaiko** [comment]
People, remember that CP will always rely on the Giants AIVehicleUtil.driveToPoint() function's speed limitations, which will always kick in during turns especially during changes in the steering angle and you'll almost always see a speed below the one you set as turn speed in the HUD.

**Comment by Petje81** [comment]
- Amazone Pantera (ingame modhub mod with only changed workspeed to 25 kph)
- Turn speed: 15kph
- Field speed: 25 kph
- Lone Oak Field 50 on a contract
- Courseplay 6 v311.2
- Created a new course
- 1 Headland on "smooth"
- "use AI turn maneuvers" deactivated
- PPC activated

Well damnnnn! This works smooth!
Headlands... same as always, it just works.

The row end turns are beautiful! It does drop speed pretty far (sometimes 2 or 3 kph for a split second) but this gives the sprayer enough time to get the wheels turned and just follow the path as intented. Right after (and i mean just smoothly right after..) the speed drop it picks up again and makes the turn smooth as butter.

I just had to try switching off PPC right before one of the last turns and the result was the same as the title says.

So my conclusion on this test:

Nice job! And it feels kinda realistic too when focussing on the sprayer only (i've worked with sprayers and this is damn near reality). No more overshooting and just one smooth turn before starting another row.

Now this made me curious of how the rest of the vehicles and machines react to this. Will test these too ofcourse.

**Comment by Voidheartd** [comment]
> How can the Issue be solved, when turning off the fix solve it ?
> Turn tests later..... yay.....

Because the new update fixed turns even with PPC off? There is no wobbling even at 15MPH now (didn't try any faster speed).

And about the speed limitations, that is why I noticed a difference. With PPC off the tractor is taking the turn at 15MPH and backing up at 10MPH (both of those being the values I have set) just fine. With PPC on the tractor goes at 7MPH through the turn and backs up at 7MPH.

**Comment by Tensuko** [comment]
Wobbling wasn't the Issue here, it was missing a waypoint and going in circles.
That was fixed by PPC and will be implemented, as it fixes more Issues.

**Comment by Voidheartd** [comment]
> Wobbling wasn't the Issue here, it was missing a waypoint and going in circles.
> That was fixed by PPC and will be implemented, as it fixes more Issues.

The wobbling and the missing waypoints seemed to be caused by the same thing for me, the tractor turning its wheels too slowly. That appears to have gone away with 311.1.

So as I said, that update fixed the problem for me even with PPC disabled. I'm still wondering why PPC is forcing a limited speed because that is not happening with it off. That makes it a worse choice if there are no problems with it off.

**Comment by pvaiko** [comment]
Also, as I said a million times: it is not PPC which is forcing a lower speed, it is the Giants physics engine slowing down the vehicle until the wheel is turning.

WIth PPC off, this physics engine function is disabled and you may get lucky if your turning radius is high enough it won't cause any issues.

**Comment by pvaiko** [comment]
Oh, and it has never been disabled before v306, so in the past year or so you were never be able to drive that fast in a turn.

**Comment by Petje81** [comment]
Yesterday i did the first really square field where i set the headland corners to "turn" instead of "smooth"  and then there was something odd going on.

On headlands same vehicle as ever, the above Amazone Pantera, stopped spraying right before the turn (a good thing), turned 90 degrees to the next headland pass, doesn't set the wheels straight and reverses some 200 degrees, goes forward with the wheels still steered until it comes back right around the place it started backing up, straightens the wheels and backs up to the edge of the field before hitting the headland nicely.
So the extra 200 degree reverse is a bit odd..
Maybe linked to [https://github.com/Courseplay/courseplay/issues/4421](url)

Must say i am still using v311.2 so i will test this with the latest version as soon as i can.

**Comment by Tensuko** [comment]
Pls test with V316+ and report, so far all turns working fine for me, but can't test all tools in all situations...

**Comment by Tensuko** [comment]
Closed due to inactivity.
Turn rework should have solved this as several tries was successful.
If have any Issue, make a new Issue with logfile with debugchannel 12 and 14 on.
